src/Accounts/Authentication/Factories/AuthenticationFactory.cs

```diff
@@ -96,79 +106,52 @@ public IServicePrincipalKeyStore KeyStore
-                {
-                    TenantId = tenant,
-                    UserId = account.Id,
-                    LoginType = AzureAccount.AccountType.AccessToken
-                };
-
-                if ((string.Equals(resourceId, environment.AzureKeyVaultServiceEndpointResourceId, StringComparison.OrdinalIgnoreCase)
-                     || string.Equals(AzureEnvironment.Endpoint.AzureKeyVaultServiceEndpointResourceId, resourceId, StringComparison.OrdinalIgnoreCase))
-                     && account.IsPropertySet(AzureAccount.Property.KeyVaultAccessToken))
-                {
-                    rawToken.AccessToken = account.GetProperty(AzureAccount.Property.KeyVaultAccessToken);
-                }
-                else if ((string.Equals(resourceId, environment.GraphEndpointResourceId, StringComparison.OrdinalIgnoreCase)
-                    || string.Equals(AzureEnvironment.Endpoint.GraphEndpointResourceId, resourceId, StringComparison.OrdinalIgnoreCase))
-                    && account.IsPropertySet(AzureAccount.Property.GraphAccessToken))
-                {
-                    rawToken.AccessToken = account.GetProperty(AzureAccount.Property.GraphAccessToken);
-                }
-                else if ((string.Equals(resourceId, environment.ActiveDirectoryServiceEndpointResourceId, StringComparison.OrdinalIgnoreCase)
-                    || string.Equals(AzureEnvironment.Endpoint.ActiveDirectoryServiceEndpointResourceId, resourceId, StringComparison.OrdinalIgnoreCase))
-                    && account.IsPropertySet(AzureAccount.Property.AccessToken))
+                try
                 {
-                    rawToken.AccessToken = account.GetAccessToken();
+                    while (processAuthenticator != null && processAuthenticator.TryAuthenticate(GetAuthenticationParameters(tokenCacheProvider, account, environment, tenant, password, promptBehavior, promptAction, tokenCache, resourceId), out authToken))
+                    {
+                        token = authToken?.ConfigureAwait(true).GetAwaiter().GetResult();
+                        if (token != null)
+                        {
+                            // token.UserId is null when getting tenant token in ADFS environment
+                            account.Id = token.UserId ?? account.Id;
+                            if (!string.IsNullOrEmpty(token.HomeAccountId))
+                            {
+                                account.SetProperty(AzureAccount.Property.HomeAccountId, token.HomeAccountId);
+                            }
+                            break;
+                        }
+
+                        processAuthenticator = processAuthenticator.Next;
+                    }
                 }
-                else
+                catch (Exception e)
                 {
-                    throw new InvalidOperationException(string.Format(Resources.AccessTokenResourceNotFound, resourceId));
+                    if (retries == 0)
```

Comment:
Should not retry if InnerException is MsalServiceException or MsalUiRequiredException